# Data Exploration

## Data Import

In [75]:
import pandas as pd
import numpy as np

In [76]:
d_hjorring = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Hjorring")
print(d_hjorring.head())

    SerialID  building_id  zone_id     SEEB  \
0  311487854            0        0  2-1-3-0   
1  311235119            0        0  2-1-3-0   
2  311235135            0        0  2-1-3-0   
3  311235144            0        0  2-1-3-0   
4  311488415            0        0  2-1-3-0   

                                    S_text  \
0        Fjernvarme uden veksler - direkte   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4        Fjernvarme uden veksler - direkte   

                                              L_text                Type  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
1  Ejendommen opvarmes med fjernvarme, og anlægge...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
4  Bygningen opvarmes med fjernvarme. Anlægget er...  Direct

In [77]:
d_favrskov = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Favrskov")
print(d_favrskov.head())

    SerialID  building_id  zone_id     SEEB  \
0  311239024            0        0  2-1-3-0   
1  311239047            0        0  2-1-3-0   
2  311239049            0        0  2-1-3-0   
3  311244995            0        0  2-1-3-0   
4  311251256            0        0  2-1-3-0   

                                    S_text  \
0  Fjernvarme uden veksler (direkte anlæg)   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4  Fjernvarme uden veksler (direkte anlæg)   

                                              L_text                Type  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
1  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
4  :Klynge A, B, C og D:\nBygningerne opvarmes me...  Direct

In [78]:
d_aarhus = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Aarhus")
print(d_aarhus.head())

    SerialID  building_id  zone_id     SEEB  \
0  311075771            0        0  2-1-3-0   
1  311075781            0        0  2-1-3-0   
2  311075799            0        0  2-1-3-0   
3  311075871            0        0  2-1-3-0   
4  311075881            0        0  2-1-3-0   

                                    S_text  \
0  Fjernvarme uden veksler (direkte anlæg)   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4                       Direkte fjernvarme   

                                              L_text                Type  \
0  Ejendommen opvarmes med fjernvarme, og anlægge...  DirectDistrictHeat   
1  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3                                                NaN  DirectDistrictHeat   
4                                                NaN  Direct

In [79]:
d_gt = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Ground Truth")
print(d_gt.head())

    SerialID  building_id  zone_id     SEEB  \
0  311117034            0        0  2-1-3-0   
1  311117893      2050872  2050872  2-1-3-0   
2  311119019            0        0  2-1-3-0   
3  311121828            0        0  2-1-3-0   
4  311122000            0        0  2-1-3-0   

                                              S_text  \
0  Fjernvarme med isoleret veksler (indirekte anl...   
1                            Eksisterende fjernvarme   
2  Fjernvarme med uisoleret veksler (indirekte an...   
3  Fjernvarme med isoleret veksler (indirekte anl...   
4  Fjernvarme med isoleret veksler (indirekte anl...   

                                              L_text  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...   
1                                                NaN   
2  Bygningen opvarmes med fjernvarme. Anlægget er...   
3  Ejendommen opvarmes med fjernvarme fra HOFOR.\...   
4  Bygningen opvarmes med fjernvarme. Anlægget er...   

                        Type  Pieces1  Manu

## Data Size

In [80]:
print("Data Favrskov shape: ", d_favrskov.shape)
print("Data Hjorring shape: ", d_hjorring.shape)
print("Data Aarhus shape: ", d_aarhus.shape)
print("Data Ground Truth shape: ", d_gt.shape)

Data Favrskov shape:  (111, 15)
Data Hjorring shape:  (6138, 15)
Data Aarhus shape:  (29542, 15)
Data Ground Truth shape:  (2272, 42)


Let's merge the dataset on the SerialID to check whether the ground truth data are present in the rest of the samples.

In [81]:
d_f_gt = pd.merge(left=d_favrskov, right=d_gt, on="SerialID")
d_h_gt = pd.merge(left=d_hjorring, right=d_gt, on="SerialID")
d_a_gt = pd.merge(left=d_aarhus, right=d_gt, on="SerialID")
print(d_f_gt.shape)
print(d_h_gt.shape)
print(d_a_gt.shape)

(0, 56)
(0, 56)
(0, 56)


From the previous result, we can say that the ground truth is not present in data.
Thus we have 2272 labeled entries.

In [82]:
print(d_favrskov.shape[0] + d_hjorring.shape[0] + d_aarhus.shape[0])

35791


On the other hand, we have 35791 entries we want to label.

## Features of the Ground Truths

In [83]:
print(d_gt[d_gt['SerialID'] == 311795931].iloc[0])

SerialID                                                                   311795931
building_id                                                                        0
zone_id                                                                            0
SEEB                                                                         2-1-3-0
S_text                                                           Fjernvarmeforsyning
L_text                             Forsyningsformen er fjernvarme som leveres af ...
Type                                                       DistrictHeatWithExchanger
Pieces1                                                                          2.0
Manufacturer1                                                                   Reci
SubType1                                                                       VT 90
HxType1                                                                   Rørveksler
NominelEffectEach1                                             30

In [84]:
d_gt_labels = d_gt.iloc[:, 7:31]
print(d_gt_labels.columns)
print("Number of lables: ", d_gt_labels.shape[1])

Index(['Pieces1', 'Manufacturer1', 'SubType1', 'HxType1', 'NominelEffectEach1',
       'Year1', 'Pieces2', 'Manufacturer2', 'SubType2', 'HxType2',
       'NominelEffectEach2', 'Year2', 'Pieces3', 'Manufacturer3', 'SubType3',
       'HxType3', 'NominelEffectEach3', 'Year3', 'Pieces4', 'Manufacturer4',
       'SubType4', 'HxType4', 'NominelEffectEach4', 'Year4'],
      dtype='object')
Number of lables:  24


In [85]:
print(d_gt_labels.isnull().mean() * 100)

Pieces1                 0.000000
Manufacturer1           0.000000
SubType1               82.394366
HxType1                 7.262324
NominelEffectEach1     92.561620
Year1                  58.846831
Pieces2                95.774648
Manufacturer2          95.730634
SubType2               96.830986
HxType2                96.082746
NominelEffectEach2     99.691901
Year2                  96.390845
Pieces3                99.647887
Manufacturer3          99.647887
SubType3               99.779930
HxType3                99.647887
NominelEffectEach3     99.911972
Year3                  99.691901
Pieces4                99.735915
Manufacturer4          99.735915
SubType4               99.823944
HxType4                99.735915
NominelEffectEach4    100.000000
Year4                  99.823944
dtype: float64


From the previous result, we know that every entry has informations about `Pieces1` and `Manufacturer1`.
We can see that `NominalEffectEach4` is always empty, and thus we do not have information about this column.
Finally, we can see that data are very sparse, indeed very few columns are filled.

In [86]:
null_counts = d_gt_labels.isnull().sum(axis=1)
missing_values = null_counts.value_counts()

missing_index = np.array(missing_values.index)
missing_index.sort()
missing_values = np.array(missing_values[missing_index])
missing_values_cum = np.cumsum(missing_values)
percentages = missing_values_cum / missing_values_cum[-1] * 100

for i in range(missing_index.shape[0]):
    print("Number of entries: ", missing_values_cum[i], f"\tCorresponding percentage: {percentages[i]:.3f} No more than {missing_index[i]} entries missing")

Number of entries:  1 	Corresponding percentage: 0.044 No more than 3 entries missing
Number of entries:  5 	Corresponding percentage: 0.220 No more than 6 entries missing
Number of entries:  6 	Corresponding percentage: 0.264 No more than 7 entries missing
Number of entries:  7 	Corresponding percentage: 0.308 No more than 8 entries missing
Number of entries:  8 	Corresponding percentage: 0.352 No more than 10 entries missing
Number of entries:  10 	Corresponding percentage: 0.440 No more than 12 entries missing
Number of entries:  61 	Corresponding percentage: 2.685 No more than 13 entries missing
Number of entries:  65 	Corresponding percentage: 2.861 No more than 14 entries missing
Number of entries:  85 	Corresponding percentage: 3.741 No more than 16 entries missing
Number of entries:  90 	Corresponding percentage: 3.961 No more than 17 entries missing
Number of entries:  164 	Corresponding percentage: 7.218 No more than 18 entries missing
Number of entries:  396 	Corresponding p

In [100]:
print(f"Average length of S_text: {d_gt['S_text'].str.len().mean():.2f}\t standard deviation: {d_gt['S_text'].str.len().std():.2f}")
print(f"Average length of L_text: {d_gt['L_text'].str.len().mean():.2f} standard deviation: {d_gt['L_text'].str.len().std():.2f}")

Average length of S_text: 48.65	 standard deviation: 14.20
Average length of L_text: 195.77 standard deviation: 175.14
